### PR TITLE
Feature/stats by processor

### LINKED ISSUE #244: A chart that shows the consent state per processor
(this might be weird, might need multiple charts, not currently sure how to get all processor states on one chart.)

### PR DESCRIPTION
**Pull request by Jakedavies**
closes #244

### CODE REVIEW

File: admin/src/containers/Dashboard/Dashboard.js [hunk 1]
```diff
@@ -1,20 +1,51 @@
...
-import ConsentDial from './Widgets/ConsentDial';
+import ControllerConsentDial from 'components/Dashboard/Widgets/ControllerConsentDial';
+import ConsentByProcessor from 'components/Dashboard/Widgets/ConsentByProcessor';
+import { StatsConsumer } from './StatsContext';
 
-class Dashboard extends Component {
+// so we can invoke the fetching from here rather than the individual component
+class Stats extends Component {
+  componentDidMount() {
+    console.log(this.props);
```
Comment (fjcero) [inline admin/src/containers/Dashboard/Dashboard.js]: 😬

File: admin/src/containers/Dashboard/StatsContext.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  setLoading = () => {
+    this.setState({ loading: true });
+  };
+
+  fetchStats() {
+    // prevents doubling up on requests that are in flight
+    if (this.state.isLoading) return this.request;
+    this.request = internalFetch(`${config.API_URL}/api/management/stats`)
+      .then(({ data }) => {
+        console.log(data);
```
Comment (fjcero) [inline admin/src/containers/Dashboard/StatsContext.js]: 😬

File: admin/src/components/core/cards/dashboard/ItemsCard.js [hunk 3]
```diff
@@ -1,19 +1,20 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 
-import Card from './Card';
-import GraphBar from './GraphBar';
+import Card from 'components/core/cards/dashboard/Card';
+import GraphBar from 'components/core/cards/dashboard/GraphBar';
 
 const ItemsCard = props => {
+  console.log(props);
```
Comment (fjcero) [inline admin/src/components/core/cards/dashboard/ItemsCard.js]: 😬

File: admin/src/components/core/cards/dashboard/ItemsCard.js [hunk 4]
```diff
@@ -25,13 +26,8 @@ const ItemsCard = props => {
...
   cols: PropTypes.number,
-  title: PropTypes.string,
-  data: PropTypes.arrayOf(
-    PropTypes.shape({
-      name: PropTypes.string,
-      consented: PropTypes.number
-    })
-  ),
+  fillPercent: PropTypes.number,
+  name: PropTypes.string,
```
Comment (fjcero) [inline admin/src/components/core/cards/dashboard/ItemsCard.js]: Shouldn’t be this 2 props part of data as an array of objects?

File: admin/src/components/Dashboard/Widgets/ConsentByProcessor.test.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+          name: 'p1',
+          consented: 1
+        }
+      ]
+    };
+    const component = shallow(<ConsentByProcessor {...props} />);
+    expect(toJson(component)).toMatchSnapshot();
+  });
+
+  it('should render if numbers are 0', () => {
```
Comment (maksymilian-majer) [inline admin/src/components/Dashboard/Widgets/ConsentByProcessor.test.js]: This looks like an exact copy of the test above. Is there a difference?

File: admin/src/components/Dashboard/Widgets/__snapshots__/ControllerConsentDial.test.js.snap [hunk 6]
```diff
@@ -7,15 +7,15 @@ exports[`(Component) ConsentDial should render 1`] = `
...
       },
       Object {
         "x": "Not Consented",
-        "y": undefined,
+        "y": 10,
       },
     ]
   }
-  text="100% of subjects have consented"
+  text="1100% of subjects have consented"
```
Comment (maksymilian-majer) [inline admin/src/components/Dashboard/Widgets/__snapshots__/ControllerConsentDial.test.js.snap]: That is obviously wrong it should be 9% (1/11), but I fixed this and it's already on master. Should be updated after rebase.

File: admin/src/containers/Dashboard/StatsContext.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import PropTypes from 'prop-types';
+import internalFetch from 'helpers/internal-fetch';
+
+const StatsContext = createContext({
+  isLoading: false,
+  stats: {},
+  fetchStats: () => {}
+});
+
+export class StatsProvider extends Component {
```
Comment (maksymilian-majer) [inline admin/src/containers/Dashboard/StatsContext.js]: Do we need a context for fetching data without interaction?

Comment (Jakedavies) [reply admin/src/containers/Dashboard/StatsContext.js]: 👌

File: admin/src/components/Dashboard/Widgets/ConsentByProcessor.test.js (hunk 5 above)
Comment (Jakedavies) [reply admin/src/components/Dashboard/Widgets/ConsentByProcessor.test.js]: I made this test actually test the 0 case now 👍

**Comment by Jakedavies** [comment]
All comments addressed and tests updated.

**Pull request merged**